extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/tenant/HibernateCurrentTenantIdentifierResolver.java

```diff
@@ -0,0 +1,89 @@
+import org.jboss.logging.Logger;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.InstanceHandle;
+import io.quarkus.hibernate.orm.runtime.JPAConfig;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Maps from the Quarkus {@link TenantResolver} to the Hibernate {@link CurrentTenantIdentifierResolver} model.
+ * 
+ * @author Michael Schnell
+ *
+ */
+public class HibernateCurrentTenantIdentifierResolver implements CurrentTenantIdentifierResolver {
+
+    private static final Logger LOG = Logger.getLogger(HibernateCurrentTenantIdentifierResolver.class);
+
+    @Override
+    public String resolveCurrentTenantIdentifier() {
+
+        TenantResolver resolver = tenantResolver();
+
+        // Make sure that we're in a request or return default tenant
+        if (!Arc.container().requestContext().isActive()) {
+            return resolver.getDefaultHibernateOrmTenantId();
+        }
+
+        RoutingContext routingContext = routingContext();
+        String tenantId = resolver.resolveHibernateOrmTenantId(routingContext);
+        if (tenantId == null) {
+            throw new IllegalStateException("Method 'CurrentTenantIdentifierResolver."
+                    + "resolveHibernateOrmTenantId(RoutingContext)' returned a null value. "
+                    + "Unfortunately Hibernate ORM does not allow null for tenant identifiers. "
+                    + "Please use a non-null value!");
+        }
+        LOG.debugv("resolveCurrentTenantIdentifier(): {0}", tenantId);
+        return tenantId;
+
+    }
+
+    @Override
+    public boolean validateExistingCurrentSessions() {
+        return jpaConfig().isValidateTenantInCurrentSessions();
+    }
+
+    /**
+     * Returns the singleton JPA configuration instance.
+     * 
+     * @return JPA configuration.
+     */
+    private static JPAConfig jpaConfig() {
+        InstanceHandle<JPAConfig> jpaConfigInstance = Arc.container().instance(JPAConfig.class);
+        if (!jpaConfigInstance.isAvailable()) {
+            throw new IllegalStateException("No instance of JPAConfig found");
+        }
+        return jpaConfigInstance.get();
+    }
+
+    /**
+     * Retrieves the routing context or fails if it is not available.
+     * 
+     * @return Current routing context.
+     */
+    private static RoutingContext routingContext() {
+        InstanceHandle<RoutingContext> routingContextInstance = Arc.container().instance(RoutingContext.class);
+        if (!routingContextInstance.isAvailable()) {
+            throw new IllegalStateException("No instance of " + RoutingContext.class.getName() + " was found.");
+        }
+        return routingContextInstance.get();
+    }
+
+    /**
+     * Retrieves the tenant resolver or fails if it is not available.
+     * 
+     * @return Current tenant resolver.
+     */
+    private static TenantResolver tenantResolver() {
+        InstanceHandle<TenantResolver> resolverInstance = Arc.container().instance(TenantResolver.class);
```

Comment:
Resolving the reference to the right `TenantResolver` via Arc is not necessary and actually a problem as we want to be able to start multiple instances of Hibernate ORM, which could be configured differently.

You should be able to simply inject the `HibernateCurrentTenantIdentifierResolver` instance into the ORM configuration; for example this should work:

    configuration.put( org.hibernate.cfg.AvailableSettings.MULTI_TENANT_IDENTIFIER_RESOLVER, new HibernateCurrentTenantIdentifierResolver( tenantResolver ) );

This way you can store the tenant into a `final` field, properly and simply initialized in its constructor.

Reply:
Which " ORM configuration" do you exactly refer to?

Reply:
It's the `cfg` Map variable you can see in `io.quarkus.hibernate.orm.runtime.boot.FastBootMetadataBuilder#mergeSettings`, were we set most other automatically inferred configuration properties.

Reply:
OK, I guess it is then not the "FastBootMetadataBuilder", but the [FastBootEntityManagerFactoryBuilder](https://github.com/michael-schnell/quarkus/blob/a24a32087d6f107eb5fe4fff8f757a1baf928850/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootEntityManagerFactoryBuilder.java#L166) as the instance of HibernateCurrentTenantIdentifierResolver is created there.

But I assume this will not work as the "FastBootEntityManagerFactoryBuilder" will be created at startup (maybe application scope), but the "TenantResolver" needs to be resolved within a request scope.

Is there really a need to support multiple instances of Hibernate ORM with multiple different implementations of the "TenantResolver"? This will get much more complicated. I guess the only way then would be to configure the class name of the actual "TenantResolver" per Hibernate ORM instance. But how do I configure multiple instances of Hibernate ORM? I haven't seen such a feature anywhere in the docs.

Reply:
> Is there really a need to support multiple instances of Hibernate ORM with multiple different implementations of the "TenantResolver"?

I'm afraid so: if we take such shortcuts it would all get confusing to use.

> But how do I configure multiple instances of Hibernate ORM? I haven't seen such a feature anywhere in the docs.

We don't have the capability yet but it is in high demand; the problem is that as lots of people keep adding features and changes that make it conceptually harder we'll never be able to complete that.. which is why I please need some help in making sure all newly merged code at least take the requirement into account.

I do understand it gets a little bit more complicated but I just want to make sure the design will make it possible at a conceptual level - not asking you to actually implement it.

Thanks!

Reply:
OK, let's assume we have multiple instances of Hibernate ORM. Then we will need to distinguish multiple instances of the tenant resolver and the resolvers may have different types. This could be achieved using CDI qualifiers and explicitly naming the resolver class.

Then the solution would have two parts:
1. Define the type of the resolver by having it's fully qualified class name in the config
2. Define a qualifier in the config that maps somehow to the Hibernate ORM instance

Example:
```
quarkus.hibernate-orm.one.multitenant=SCHEMA
quarkus.hibernate-orm.one.multitenancy-resolver=a.b.c.MyResolver // Implements TenantResolver
quarkus.hibernate-orm.one.multitenancy-qualifier=one // Qualifier
quarkus.hibernate-orm.two.multitenant=DATABASE
quarkus.hibernate-orm.two.multitenancy-resolver=a.b.c.OtherResolver // Implements TenantResolver
quarkus.hibernate-orm.two.multitenancy-qualifier=two // Qualifier
```
Currently we don't have multiple instances and we don't want to bother the user with such edge case configurations. The "95% case" will simply be one instance. Therefore we need a way to keep it still simple. This is exactly how it is implemented now and it is open to above future design.

In the FastBootEntityManagerFactoryBuilder the HibernateCurrentTenantIdentifierResolver instance is created and might use some configuration options later on, that are not needed now:
```java
if (strategy != null && strategy != MultiTenancyStrategy.NONE) {
    options.applyMultiTenancyStrategy(strategy);
    options.applyCurrentTenantIdentifierResolver(new HibernateCurrentTenantIdentifierResolver());
}
```
In the HibernateCurrentTenantIdentifierResolver class we still resolve the TenantResolver dynamically (using Arc), because it is RequestScoped and we cannot create it at boot time in the FastBootEntityManagerFactoryBuilder:
```java
private static TenantResolver tenantResolver() {
    InstanceHandle<TenantResolver> resolverInstance = Arc.container().instance(TenantResolver.class);
    ....
}
```
Later on, when multiple Hibernate ORM instances are available you could use the qualifier and class from the configuration similar to this one:
```java
private static TenantResolver tenantResolver(Class<TenantResolver> resolverClass, String qualifierName) {
    InstanceHandle<TenantResolver> resolverInstance = Arc.container().instance(resolverClass,
            new AnnotationLiteral() {
                public String value() {
                    return qualifierName;
                };
            });
      ...
```

**Summary**
If we do the changes you proposed in regard to the vertx module dependency and make the TenantResolver request scoped (as suggested above), the HibernateCurrentTenantIdentifierResolver can stay as it is now.

Reply:
> Currently we don't have multiple instances and we don't want to bother the user with such edge case configurations. The "95% case" will simply be one instance. Therefore we need a way to keep it still simple. This is exactly how it is implemented now and it is open to above future design.

I like what you're saying here and totally agree that we should keep it simple for the 95% of users.

Referring to your example:

I might be missing something since I've never used CDI, but wouldn't we conceptually need the user to just specify _either_ the qualifier _or_ the class? I would expect it to be possible to only need to express a "single choice".

Further more, consider that via the Jandex index we have a full analysis of the application; this implies that we can enumerate the implementations of, for example, a TenantResolver implementation and choose it as default when only a single implementation is found.
(should we go with a single TenantResolver implementation and use qualifiers instead - assuming it's possible to use only one approach as mention in the previous paragraph - then resolving the default bean rather than using qualifiers should come even more natural?)

That would be sweet as it means the components could be bound correctly without any configuration at all; requiring configuration only for multiple persistence units which require to differentiate their settings.

@manovotn / @mkouba would you have some time to review and correct my ramblings about ArC? I'd really hope we could keep this both flexible and straight forward to use :)

Thanks!

Reply:
I have quite limited understanding of what you're trying to achieve and might need some more context, but broadly speaking - if you want to have multiple implementations of, say, `TenantResolver`, then you will end up with ambiguity if you try to just inject it based on type `TenantResolver`. You'd have to either use qualifiers, or resolve it based on the implementing class.

As for simple cases and having one `TenantResolver` by default, that's doable, you just have to have a singular existing bean with that type and without CDI qualifiers. From the above comments I understood that this is something you already have.

If the impl of `TenantResolver` is something that user defines, then they can define the qualifiers used with it as well. And if you then need to retrieve all impls of `TenantResolver` via CDI (regardless of qualifiers) you can do that too with (for instance) `Arc.container().instance(TenantResolver.class, Any.Literal.INSTANCE)` and then iterate over them.

Reply:
If I am blabbering about something else then what you wanted to know, then please give me some more info and I'll try to help some more :-)

Reply:
At the end I just wanted to show that using "Arc.container().instance(..)" is not "evil" here. It is required, because we need to resolve a request scoped bean and therefore we cannot do it at boot time. In addition to that it allows us in a later stage to change this as needed in HibernateCurrentTenantIdentifierResolver. But no need to do it now as we don't have multiple Hibernate ORM instances yet. So no need for qualifiers and other stuff now, but a way to go in the future.

I already tested the new classes with the new Hibernate ORM multitenancy quickstart and it still works finr with the changes above. I will now commit them to continue here with the next review steps.